mkosi/distributions/gentoo.py

```diff
@@ -295,41 +220,60 @@ def mkosi_conf(self) -> None:
-        invoke_emerge(self.state, opts=self.EMERGE_UPDATE_OPTS, pkgs=self.pkgs['sys'])
-
-    def depclean(self) -> None:
-        invoke_emerge(self.state, actions=["--depclean"])
-
-
-class GentooInstaller(DistributionInstaller):
-    @classmethod
-    def filesystem(cls) -> str:
-        return "ext4"
-
-    @staticmethod
-    def kernel_image(name: str, architecture: Architecture) -> Path:
-        kimg_path = {
-            Architecture.x86_64: "arch/x86/boot/bzImage",
-            Architecture.arm64: "arch/arm64/boot/Image.gz",
-            Architecture.arm: "arch/arm/boot/zImage",
-        }[architecture]
-
-        return Path(f"usr/src/linux-{name}") / kimg_path
-
-    @classmethod
-    def install(cls, state: MkosiState) -> None:
-        Gentoo(state)
+        repos_cfg = cls.stage3_cache / BINREPOS_CONF_FILE
+        with repos_cfg.open(mode='a') as f:
+            for repo in state.config.repositories:
+                f.write(
+                    dedent(
+                        f"""\
+                        # MKOSI
+                        [binhost]
+                        sync-uri = {repo}
+                        """
+                    )
+                )
+
+        root_portage_cfg = state.root
+        root_portage_cfg /= user_config_path.relative_to(cls.stage3_cache)
+        root_portage_cfg.mkdir(parents=True, exist_ok=True)
+        copy_path(user_config_path, root_portage_cfg)
+
+        bwrap_params: list[PathString] = [
+            "--bind", state.cache_dir / "repos", "/var/db/repos"
+        ]
+        run_workspace_command(cls.stage3_cache, ["/usr/bin/emerge-webrsync"],
+                              bwrap_params=bwrap_params, network=True)
+
+        opts = [
+            "--with-bdeps=n",
+            "--complete-graph-if-new-use=y",
+            "--verbose-conflicts",
+            "--changed-use",
+            "--newuse",
+        ]
+        env = {}
+        env.update(emerge_vars)
+        env.update({"USE": f"{emerge_vars['USE']} build"})
+        with complete_step("merging stage2"):
+            invoke_emerge(state, sysroot=cls.stage3_cache, opts=opts+["--emptytree", "--nodeps"],
+                          pkgs=["sys-apps/baselayout", "sys-apps/util-linux"], env=env)
+        opts += ["--noreplace", "--root-deps=rdeps"]
+        with complete_step("Merging bare minimal atoms"):
+            invoke_emerge(state, sysroot=cls.stage3_cache, opts=opts+["--exclude", "sys-devel/*"],
+                          pkgs=["app-shells/bash", "sys-apps/systemd"], env=emerge_vars)
+        with complete_step("Merging atoms required for boot"):
+            invoke_emerge(state, sysroot=cls.stage3_cache, opts=opts,
+                          pkgs=["sys-kernel/gentoo-kernel-bin"])
+        if state.config.make_initrd:
+            return
+
+        invoke_emerge(state, sysroot=cls.stage3_cache, opts=opts,
+                      pkgs=["@system"])
```

Comment:
We should not install `@system` by default. For all other distributions, we only install the package providing the base filesystem layout and the package providing /etc/passwd. We should do the same for gentoo.

Reply:
Yes please remove it. By default we try to build the most minimal image possible, even if it's not useful because it's much easier to add missing stuff than it is to remove one of mkosi's defaults that a user doesn't need.

Think of use cases like building sysexts or portable services where systemd is not necessary in the image.